paho/client.go

```diff
@@ -551,6 +596,73 @@ func (c *Client) close() {
 	c.debug.Println("acks tracker reset")
 }
 
+func (c *Client) readLoop() {
+	c.debug.Println("readLoop starting")
+
+	incoming := make(chan *packets.ControlPacket)
+	readNext := func() {
+		c.debug.Println("reading next packet")
+		packet, err := packets.ReadPacket(c.Conn)
+		if err != nil {
+			c.errors.Printf("read packet error: %v\n", err)
+			go c.error(err)
+			return
+		}
+
+		packetIdString := ""
+		if id := packet.PacketID(); id != 0 {
+			packetIdString = fmt.Sprintf(" with id=%d", id)
+		}
+		c.debug.Printf("received %s%s\n", packet.PacketType(), packetIdString)
+
+		incoming <- packet
```

Comment:
goroutine leak here.

`c.stop` is closed, the outer `for` exits so `incoming <- packet` will block (as nothing is receiving on `incoming`).

I don't really see the benefit of having using the extra goroutines here `packets.ReadPacket(c.Conn)` will return an error when `c.Conn` is closed (so `readLoop()` can be simplified to a few lines).

Reply:
agree